src/coreclr/inc/clrhost.h

```diff
@@ -29,16 +29,29 @@ using std::nothrow;
 #define _DEBUG_IMPL 1
 #endif
 
+struct PreserveLastErrorHolder
+{
+    PreserveLastErrorHolder()
+    {
+        m_dwLastError = ::GetLastError();
+    }
+
+    ~PreserveLastErrorHolder()
+    {
+        ::SetLastError(m_dwLastError);
+    }
+private:
+    DWORD m_dwLastError;
+};
+
+
 #define BEGIN_PRESERVE_LAST_ERROR \
     { \
-        DWORD __dwLastError = ::GetLastError(); \
-        DEBUG_ASSURE_NO_RETURN_BEGIN(PRESERVE_LAST_ERROR); \
-            {
+        PreserveLastErrorHolder __preserveLastErrorHolder; \
```

Comment:
This is going to introduce exception handling in the methods where it is used. Is it ok to do that in all places where this macro is used?